File: labs/control-ui/api-shadow/mobile_engine_provider.ts
export const EngineState = {
  Stopped: 'Stopped',
  Starting: 'Starting',
  Running: 'Running',
  Degraded: 'Degraded',
  Stopping: 'Stopping',
  Error: 'Error',
} as const;
export type EngineState = (typeof EngineState)[keyof typeof EngineState];

export interface EngineMetrics {
  uptimeSecs: number;
  rxBytes: number;
  txBytes: number;
  activeTunnels: number;
  lastHeartbeat: number;
}

export interface EngineConfig {
  bindAddress: string;
  socks5Port: number;
  httpPort: number;
  dnsPort: number;
  mtu: number;
  enableIpv6: boolean;
}

export class MobileEngineProvider {
  private state: EngineState = EngineState.Stopped;
  private metrics: EngineMetrics = {
    uptimeSecs: 0,
    rxBytes: 0,
    txBytes: 0,
    activeTunnels: 0,
    lastHeartbeat: 0,
  };
  private startTimestamp: number = 0;
  public config: EngineConfig;
  constructor(config: EngineConfig) {
    this.config = config;
  }

  startEngine(timestamp: number): void {
    if (this.state === EngineState.Running) {
      throw new Error('Engine is already running');
    }
    if (this.config.socks5Port === 0 || this.config.httpPort === 0) {
      this.state = EngineState.Error;
      throw new Error('Invalid port configuration');
    }

    this.state = EngineState.Starting;
    this.startTimestamp = timestamp;
    this.metrics.lastHeartbeat = timestamp;
    this.metrics.activeTunnels = 1;
    this.state = EngineState.Running;
  }

  stopEngine(): void {
    if (this.state === EngineState.Stopped) {
      return;
    }
    this.state = EngineState.Stopping;
    this.metrics.activeTunnels = 0;
    this.state = EngineState.Stopped;
  }

  recordTraffic(rx: number, tx: number, timestamp: number): void {
    this.metrics.rxBytes += rx;
    this.metrics.txBytes += tx;
    this.metrics.lastHeartbeat = timestamp;
    if (this.startTimestamp > 0 && timestamp >= this.startTimestamp) {
      this.metrics.uptimeSecs = timestamp - this.startTimestamp;
    }
  }

  markDegraded(degraded: boolean): void {
    if (this.state === EngineState.Running && degraded) {
      this.state = EngineState.Degraded;
    } else if (this.state === EngineState.Degraded && !degraded) {
      this.state = EngineState.Running;
    }
  }

  getState(): EngineState {
    return this.state;
  }

  getMetrics(): EngineMetrics {
    return { ...this.metrics };
  }
}

File: labs/mobile/donor-archive/tunnel/MobileEngineProvider.kt
package com.luminet.android.tunnel

enum class MobileEngineState {
    STOPPED,
    STARTING,
    RUNNING,
    DEGRADED,
    STOPPING,
    ERROR
}

data class MobileEngineMetrics(
    var uptimeSecs: Long = 0,
    var rxBytes: Long = 0,
    var txBytes: Long = 0,
    var activeTunnels: Int = 0,
    var lastHeartbeat: Long = 0
)

data class MobileEngineConfig(
    val bindAddress: String = "127.0.0.1",
    val socks5Port: Int = 10808,
    val httpPort: Int = 10809,
    val dnsPort: Int = 5353,
    val mtu: Int = 1500,
    val enableIPv6: Boolean = false
)

class MobileEngineProvider(val config: MobileEngineConfig) {
    var state: MobileEngineState = MobileEngineState.STOPPED
        private set
    val metrics = MobileEngineMetrics()
    private var startTime: Long = 0

    fun startEngine(timestamp: Long) {
        if (state == MobileEngineState.RUNNING) {
            throw IllegalStateException("Engine already running")
        }
        if (config.socks5Port <= 0 || config.httpPort <= 0) {
            state = MobileEngineState.ERROR
            throw IllegalArgumentException("Invalid port configuration")
        }

        state = MobileEngineState.STARTING
        startTime = timestamp
        metrics.lastHeartbeat = timestamp
        metrics.activeTunnels = 1
        state = MobileEngineState.RUNNING
    }

    fun stopEngine() {
        if (state == MobileEngineState.STOPPED) return
        state = MobileEngineState.STOPPING
        metrics.activeTunnels = 0
        state = MobileEngineState.STOPPED
    }

    fun recordTraffic(rx: Long, tx: Long, timestamp: Long) {
        metrics.rxBytes += rx
        metrics.txBytes += tx
        metrics.lastHeartbeat = timestamp
        if (startTime > 0 && timestamp >= startTime) {
            metrics.uptimeSecs = timestamp - startTime
        }
    }

    fun markDegraded(degraded: Boolean) {
        if (state == MobileEngineState.RUNNING && degraded) {
            state = MobileEngineState.DEGRADED
        } else if (state == MobileEngineState.DEGRADED && !degraded) {
            state = MobileEngineState.RUNNING
        }
    }
}
